### ISSUE TITLE
AsSplitQuery checks ids, what causes missing data. 

### ISSUE THREAD

**Issue by timaiv** [dotnet/efcore#29430]
**Problem**
After adding `AsSplitQuery()`, collection properties are not returned or returned only for some entities.

**Reason**
After a lot of debugging time (and manual data changing in db on breakpoint), I think that if at least one id property changed between first and other requests (what happens in our realtime indexer system), this happen.

`PopulateSplitIncludeCollectionAsync` [link](https://github.com/dotnet/efcore/blob/f7333f1afa250ed33d3db52f0b531910b23a55b2/src/EFCore.Relational/Query/RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.cs#L1428)

```cs
if (!CompareIdentifiers(
                            identifierValueComparers,
                            splitQueryCollectionContext.ParentIdentifier, childIdentifier(queryContext, dbDataReader)))
                        {
                            dataReaderContext.HasNext = true;

                            return;
                        }
```

The only solution for me is to remove the `AsSplitQuery` usage and load the secondary data manually.

It's strange to me how the current mechanism works, which repeats all WHERE checks in the database, instead of using ids from the first request.
____
Unfortunately, it was very difficult to debug this problem.

**Comment by timaiv** [comment]
it is interesting that in my code, I am loading entities only where at least one collection item must be returned (and use this data further), but actually this bug may returns empty collection.
`.Where(nft=>nft.sources.First().Uri!=null)`

**Comment by roji** [comment]
@timaiv this sort of concurrency/consistency issue is one main reason why split queries aren't the default, but rather single query.

You should be able to mitigate this problem but still use split queries by using the repeatable read transaction isolation level (or snapshot on SQL Server, or serializable), but that can have additional perf costs due to increased locking etc. (carefully read your database documentation on these isolation levels).

**Comment by timaiv** [comment]
> @timaiv this sort of concurrency/consistency issue is one main reason why split queries aren't the default, but rather single query.
> 
> You should be able to mitigate this problem but still use split queries by using the repeatable read transaction isolation level (or snapshot on SQL Server, or serializable), but that can have additional perf costs due to increased locking etc. (carefully read your database documentation on these isolation levels).

Hi!
Let's say between requests value of the id property changed from 1 to 2.
It would be absolutely normal for me to see 1 (old) or  2 (new).
But not an empty collection, considering that I'm making a request in which I specify that the collection physically cannot be empty in result. `(.Where(nft=>nft.sources.First().Uri!=null))`
The documentation does not say anything about such nuances.

So i used next approach without lock.

```cs
if (itemsX.Any()) {
	var xIds = itemsX.Select(item => item.Id);
	
	_ = contextX.itemsY.Where(itemY => xIds.Contains(itemY.ItemId2))
        .ToList();
						
	_ = contextX.itemsZ.Where(itemZ => xIds.Contains(itemZ.ItemId2))
        .ToList();
}
```

**Comment by roji** [comment]
> Let's say between requests value of the id property changed from 1 to 2.
> It would be absolutely normal for me to see 1 (old) or 2 (new).
> But not an empty collection, considering that I'm making a request in which I specify that the collection physically cannot be empty in result. (.Where(nft=>nft.sources.First().Uri!=null))

I understand that this is surprising and seems problematic, but it really is the expected behavior. The point of split query is to use multiple queries to load related data. If an ID property changes between the two queries, then the 2nd query - which loads the related entities - won't find any related data. There's nothing EF can do about this, and the only way to protect against this is via a higher transaction isolation level which guarantees a consistent view of the database across the entire transaction.

> The documentation does not say anything about such nuances.

See the following note in the [docs](https://learn.microsoft.com/en-us/ef/core/querying/single-split-queries):

> While most databases guarantee data consistency for single queries, no such guarantees exist for multiple queries. If the database is updated concurrently when executing your queries, resulting data may not be consistent. You can mitigate it by wrapping the queries in a serializable or snapshot transaction, although doing so may create performance issues of its own. For more information, see your database's documentation.

To summarize, I don't see what else EF could do to mitigate this (am open to hear any ideas), and it does seem properly covered in our documentation...
